src/dstack/_internal/server/services/jobs/__init__.py

```diff
@@ -169,19 +118,138 @@ def _get_job_configurator(run_spec: RunSpec) -> JobConfigurator:
-        user=job_submission.job_provisioning_data.username,
+        hostname=jpd.hostname,
+        ssh_port=jpd.ssh_port,
+        user=jpd.username,
         ports=ports,
         id_rsa=server_ssh_private_key,
-        ssh_proxy=job_submission.job_provisioning_data.ssh_proxy,
+        ssh_proxy=jpd.ssh_proxy,
     ):
         runner_client = client.RunnerClient(port=ports[client.REMOTE_RUNNER_PORT])
-        logger.debug("Stopping runner %s", job_submission.job_provisioning_data.hostname)
         runner_client.stop()
+
+
+async def process_terminating_job(session: AsyncSession, job_model: JobModel):
+    """
+    Used by both process_terminating_jobs and process_terminating_run.
+    Caller must acquire the lock on the job.
+    """
+    if (
+        job_model.remove_at is not None
+        and job_model.remove_at.replace(tzinfo=datetime.timezone.utc) > get_current_datetime()
+    ):
+        # it's too early to terminate the instance
+        return
+
+    res = await session.execute(
+        sa.select(InstanceModel)
+        .where(
+            InstanceModel.project_id == job_model.project_id,
+            InstanceModel.job_id == job_model.id,
+        )
+        .options(sa.orm.joinedload(InstanceModel.project))
+    )
+    instance: Optional[InstanceModel] = res.scalar()
+
+    if instance is not None:
+        await wait_to_lock(PROCESSING_POOL_LOCK, PROCESSING_POOL_IDS, instance.id)
+        await session.refresh(instance)
+        try:
+            # there is an associated instance to empty
+            jpd: Optional[JobProvisioningData] = None
+            if job_model.job_provisioning_data is not None:
+                jpd = JobProvisioningData.parse_raw(job_model.job_provisioning_data)
+                logger.debug("%s: stopping container", fmt(job_model))
+                await stop_container(job_model, jpd, instance.project.ssh_private_key)
+
+            if instance.status == InstanceStatus.BUSY:
+                instance.status = InstanceStatus.IDLE
+            elif instance.status != InstanceStatus.TERMINATED:
+                # instance was CREATING or STARTING (specially for the job)
+                # schedule for termination
+                instance.status = InstanceStatus.TERMINATING
+
+            if jpd is None or not jpd.dockerized:
+                # do not reuse vastai/k8s instances
+                instance.status = InstanceStatus.TERMINATING
+
+            instance.job_id = None
+            instance.last_job_processed_at = get_current_datetime()
+            logger.info(
+                "%s: instance '%s' has been released, new status is %s",
+                fmt(job_model),
+                instance.name,
+                instance.status.name,
+            )
+            # TODO(egor-s): unregister service replica
+        finally:
+            PROCESSING_POOL_IDS.remove(instance.id)
+
+    job_model.status = job_termination_reason_to_status(job_model.termination_reason)
```

Comment:
job_model.termination_reason can be None. I understand the code assumes that if the status is Terminating, then job_model.termination_reason != None, but this fragile. Maybe handle the None case explicitly, e.g. having default termination reason.